desc/magnetic_fields/_current_potential.py

```diff
@@ -696,3 +897,662 @@ def nfp_loop(j, f):
+                / mu_0
+            )
+    else:
+        G_ext = 0
+
+    # G needed by surface current is the total G minus the external contribution
+    G = G_tot - G_ext
+    # calculate I, net toroidal current on winding surface
+    if current_helicity == 0:  # modular coils
+        I = 0
+    else:  # helical coils
+        I = G / current_helicity / eq.NFP
+
+    def B_from_K_SV(phi_mn):
+        """B from single value part of K from REGCOIL eqn 4."""
+        params = current_potential_field.params_dict
+        params["Phi_mn"] = phi_mn
+        params["I"] = 0
+        params["G"] = 0
+        Bn, _ = current_potential_field.compute_Bnormal(
+            eq.surface, eval_grid=eval_grid, source_grid=source_grid, params=params
+        )
+        return Bn
+
+    def B_from_K_secular(I, G):
+        """B from secular part of K, i.e. B^GI_{normal} from REGCOIL eqn 4."""
+        params = current_potential_field.params_dict
+        params["I"] = I
+        params["G"] = G
+        params["Phi_mn"] = jnp.zeros_like(params["Phi_mn"])
+
+        Bn, _ = current_potential_field.compute_Bnormal(
+            eq.surface, eval_grid=eval_grid, source_grid=source_grid, params=params
+        )
+        return Bn
+
+    timer = Timer()
+    # calculate the Jacobian matrix A for  Bn_SV = A*Phi_mn
+    timer.start("Jacobian Calculation")
+    A = Derivative(B_from_K_SV).compute(current_potential_field.Phi_mn)
+    timer.stop("Jacobian Calculation")
+    if verbose > 1:
+        timer.disp("Jacobian Calculation")
+
+    current_potential_field.I = float(I)
+    current_potential_field.G = float(G)
+
+    # find the normal field from the secular part of the current potential
+    B_GI_normal = B_from_K_secular(I, G)
+    if not vacuum:
+        Bn_plasma = compute_B_plasma(eq, eval_grid, source_grid, normal_only=True)
+    else:
+        Bn_plasma = jnp.zeros_like(
+            B_GI_normal
+        )  # from plasma current, currently assume is 0
+    # find external field's Bnormal contribution
+    if external_field:
+        Bn_ext, _ = external_field.compute_Bnormal(
+            eq.surface, eval_grid=eval_grid, source_grid=external_field_grid
+        )
+    else:
+        Bn_ext = jnp.zeros_like(B_GI_normal)
+
+    rhs = -(Bn_plasma + Bn_ext + B_GI_normal).T @ A
+    alphas = np.atleast_1d(alpha)
+    scan = alphas.size > 1
+
+    chi2Bs = []
+    chi2Ks = []
+    K_mags = []
+    phi_mns = []
+    Bn_arrs = []
+    fields = []
+
+    # calculate the Phi_mn which minimizes (chi^2_B + alpha*chi^2_K) for each alpha
+    for alpha in alphas:
+        printstring = f"Calculating Phi_SV for alpha = {alpha:1.5e}"
+        if verbose > 0:
+            print(
+                "#" * len(printstring)
+                + "\n"
+                + printstring
+                + "\n"
+                + "#" * len(printstring)
+            )
+
+        # calculate Phi_mn with SVD
+        Ainv_full, _ = svd_inv_null(A.T @ A + alpha * jnp.eye(A.shape[1]))
```

Comment:
check regularization to make sure it is correct

Reply:
I believe this is correct, yes. the regularization comes in when we minimize $\chi^2_B + \alpha \chi^2_K$ and the math works out to what I have here

Reply:
Can you post the math? (would also be good to include in the paper), it's not immediately obvious to me.